### PR TITLE
Recurrent DQN: Training recurrent policies

### LINKED ISSUE #2349: 💡 [REQUEST] - Port TorchRL `Recurrent DQN` tutorial from pytorch.org/rl to pytorch.org/tutorials
### 🚀 Descirbe the improvement or the new tutorial

For historical reasons, TorchRL privately hosts a bunch of tutorials.
We'd like to bring the most significant ones to pytorch tutorials for more visibility.

Here is the [tutorial](https://github.com/pytorch/rl/blob/main/tutorials/sphinx-tutorials/dqn_with_rnn.py).
In RL, we often add a RNN to a model to account for past observations when executing a policy. This of it as this: if your policy just sees a single image when playing a computer game, it will have little context about what is really happening there. If you keep a memory of past events, your performance will drastically improve. 
This is useful not only in the context of Partially Observable MDPs but more broadly than that.

Storing recurrent values can be tricky, and torchrl brings its own solution to this problem. This tutorial explains this.

Steps:
1. Port the tutorial from the RL repo to the tutorials repo.
2. Fix any formatting issues or typos.
3. Make sure the tutorial follows the tutorial template ([template_tutorial.py](https://github.com/pytorch/tutorials/blob/main/beginner_source/template_tutorial.py))
4. Preserve the original author

### Existing tutorials on this topic

_No response_

### Additional context

The tutorial should not require extra dependencies beyond those already present in requirements.txt.

cc @nairbv @sekyondaMeta @svekars @carljparker @NicolasHug @kit1980 @subramen

### PR DESCRIPTION
**Pull request by markstur**
Fixes #2349 

## Description
Add tutorial from TorchRL

## Checklist

- [x] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [x] Only one issue is addressed in this pull request
- [x] Labels from the issue that this PR is fixing are added to this pull request
- [x] No unnecessary issues are included into this pull request.

cc @vmoens @nairbv @sekyondaMeta @svekars @carljparker @NicolasHug @kit1980 @subramen

### CODE REVIEW

**Comment by svekars** [comment]
Make sure to fix the [spellcheck](https://github.com/pytorch/tutorials/actions/runs/6751205025/job/18354890295?pr=2643). Some words can be added to the en-wordlist.txt to be skipped.

File: intermediate_source/dqn_with_rnn_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+# Memory-based policies are crucial not only when the observations are partially
+# observable but also when the time dimension must be taken into account to
+# make informed decisions.
+#
+# Recurrent neural network have long been a popular tool for memory-based
+# policies. The idea is to keep a recurrent state in memory between two
+# consecutive steps, and use this as an input to the policy along with the
+# current observation.
+#
+# This tutorial shows how to incorporate an RNN in a policy.
```
Comment (vmoens) [inline intermediate_source/dqn_with_rnn_tutorial.py:36]:
```suggestion
# This tutorial shows how to incorporate an RNN in a policy using TorchRL.
```

File: intermediate_source/dqn_with_rnn_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+# When the :func:`torchrl.envs.step_mdp` function is called, the recurrent states
+# from the next state are brought to the current TensorDict. Let's see how this
+# is implemented in practice.
+
+######################################################################
+# If you are running this in Google Colab, make sure you install the following dependencies:
+#
+# .. code-block:: bash
+#
+#    !pip3 install torchrl-nightly
```
Comment (vmoens) [inline intermediate_source/dqn_with_rnn_tutorial.py:63]:
```suggestion
#    !pip3 install torchrl
```

File: intermediate_source/dqn_with_rnn_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+# output vector:
+#
+n_cells = feature(env.reset())["embed"].shape[-1]
+
+######################################################################
+# LSTM Module
+# ~~~~~~~~~~~
+#
+# TorchRL provides a specialized :class:`torchrl.modules.LSTMModule` class
+# to incorporate LSTMs in your code-base. It is a :class:`tensordict.nn.TensorDictModuleBase`
```
Comment (vmoens) [inline intermediate_source/dqn_with_rnn_tutorial.py:193]:
```suggestion
# to incorporate LSTMs in your code-base. It is a :class:`~tensordict.nn.TensorDictModuleBase`
```

File: intermediate_source/dqn_with_rnn_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+            "the output of the QValueModule (action_values, action and chosen_action_value),"
+            "the 'is_init' key that will tell us if a step is initial or not, and the "
+            "recurrent_state keys.\n",
+            data,
+        )
+    pbar.update(data.numel())
+    # it is important to pass data that is not flattened
+    rb.extend(data.unsqueeze(0).to_tensordict().cpu())
+    for _ in range(utd):
+        s = rb.sample().to(device)
```
Comment (vmoens) [inline intermediate_source/dqn_with_rnn_tutorial.py:397]:
```suggestion
        s = rb.sample().to(device, non_blocking=True)
```

File: intermediate_source/dqn_with_rnn_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+    plt.title("Test trajectory lengths")
+
+######################################################################
+# Conclusion
+# ----------
+#
+# We have seen how an RNN can be incorporated in a policy in torchrl.
+# You should now be able:
+#
+# - Create an LSTM module that acts as a :class:`tensordict.nn.TensorDictModule`
```
Comment (vmoens) [inline intermediate_source/dqn_with_rnn_tutorial.py:430]:
```suggestion
# - Create an LSTM module that acts as a :class:`~tensordict.nn.TensorDictModule`
```

File: requirements.txt [hunk 6]
```diff
@@ -25,7 +25,7 @@ tensorboard
 jinja2==3.0.3
 pytorch-lightning
 torchx
-torchrl==0.2.0
+torchrl==0.2.1
```
Comment (vmoens) [inline requirements.txt:28]:
```suggestion
torchrl==0.2.0
```
we're sticking to 0.2.0 for now, or we need to upgrade both rl and tensordict to 0.2.1

**Review by vmoens — COMMENTED** [review]
Thanks a lot for this!
I left some suggestions, mostly to get better looking links to the doc

File: requirements.txt (hunk 6 above)
Comment (markstur) [reply requirements.txt:28]: Does this work on MacOS with 0.2.1?

Comment (vmoens) [reply requirements.txt:28]: yes it should so to me we can update both dependencies

**Comment by markstur** [comment]
Thanks @vmoens , I think I got all the fixes in.  Note: I had to dev tools -> empty cache and hard reload in my browser to see a line in the chart in the preview of the tutorial.

**Review by vmoens — APPROVED** [review]
LGTM, it looks awesome! Thanks a mil!

**Review by svekars — APPROVED** [review]
